gocse/kms/provider.go

```diff
@@ -0,0 +1,271 @@
+package kms
+
+import (
+	"log"
+	"os"
+
+	"github.com/fatih/structs"
+)
+
+// GetCheckedEnv gets the specified environmental variable value and ensures
+// it isn't empty
+func GetCheckedEnv(env string) string {
+	val := os.Getenv(env)
+	if val == "" {
+		log.Fatalf("You are attempting to use a KMS provider but not providing the required environmental value for %s. Please see the README", env)
+	}
+	return val
+}
+
+// Provider is a common interface for each KMS provider (aws, azure, gcp, local)
+type Provider interface {
+	Name() string
+	Credentials() map[string]map[string]interface{}
+	DataKeyOpts() interface{}
+}
+
+// gcpKMSCredentials used to access this KMS provider
+// See https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#kmsproviders
+type gcpKMSCredentials struct {
+	Email      string `structs:"email"`
+	PrivateKey string `structs:"privateKey"`
+	Endpoint   string `structs:"endpoint,omitempty"` // optional, defaults to oauth2.googleapis.com
+}
+
+// gcpKMSDataKeyOpts are the data key options used for this KMS provider.
+// See https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#datakeyopts
+type gcpKMSDataKeyOpts struct {
+	ProjectID  string `bson:"projectId"`
+	Location   string `bson:"location"`
+	KeyRing    string `bson:"keyRing"`
+	KeyName    string `bson:"keyName"`
+	KeyVersion string `bson:"keyVersion,omitempty"` // optional, defaults to the key's primary version
+	Endpoint   string `bson:"endpoint,omitempty"`   // optional, defaults to cloudkms.googleapis.com
+}
+
+// GCP holds the credentials and master key information to use this KMS
+// Get an instance of this with the kms.GCPProvider() method
+type GCP struct {
+	credentials gcpKMSCredentials `structs:"gcp"`
+	dataKeyOpts gcpKMSDataKeyOpts
+	name        string
+}
+
+// GCPProvider reads in the required environment variables for credentials and master key
+// location to use GCP as a KMS
+func GCPProvider() *GCP {
+	// gcp kms privider credentials
+	gcpEmail := GetCheckedEnv("FLE_GCP_EMAIL")
+	gcpPrivateKey := GetCheckedEnv("FLE_GCP_PRIVATE_KEY")
+
+	// gcp data key opts
+	gcpProjectID := GetCheckedEnv("FLE_GCP_PROJ_ID")
+	gcpLocation := GetCheckedEnv("FLE_GCP_LOCATION")
+	gcpKeyRing := GetCheckedEnv("FLE_GCP_KEY_RING")
+	gcpKeyName := GetCheckedEnv("FLE_GCP_KEY_NAME")
+
+	return &GCP{
+		credentials: gcpKMSCredentials{
```

Comment:
For `credentials` and `dataKeyOpts`, I'd find it more readable if the struct field names were declared too. This would avoid the weird `""` field declarations as well:

```
credentials: gcpKMSCredentials{
     Email:          gcpEmail,
     PrivateKey: gcpPrivateKey,
}
...
```